winbuild/dir.cpp

```diff
@@ -91,23 +62,158 @@ class WinDirHandle : public watchman_dir_handle {
+    }
+    else
+    {
+      if (!FindNextFileW(hDirFind_, &findFileData)) {
+        if (GetLastError() == ERROR_NO_MORE_FILES) {
+          FindClose(hDirFind_);
+          return nullptr;
+        }
+
+        throw std::system_error(
+          GetLastError(),
+          std::system_category(),
+          "FindNextFileW");
+      }
+    }
+
+    DWORD len = WideCharToMultiByte(
+      CP_UTF8,
+      0,
+      findFileData.cFileName,
+      -1,
+      nameBuf_,
+      sizeof(nameBuf_) - 1,
+      NULL,
+      NULL);
+
+    if (len <= 0) {
+      throw std::system_error(
+        GetLastError(), std::system_category(), "WideCharToMultiByte");
+    }
+
+    nameBuf_[len] = 0;
+
+    // Populate stat info to speed up the crawler() routine
+    FILETIME_to_timespec(&findFileData.ftCreationTime, &ent_.stat.ctime);
+    FILETIME_to_timespec(&findFileData.ftLastAccessTime, &ent_.stat.atime);
+    FILETIME_to_timespec(&findFileData.ftLastWriteTime, &ent_.stat.mtime);
+
+    LARGE_INTEGER fileSize;
+    fileSize.HighPart = findFileData.nFileSizeHigh;
+    fileSize.LowPart = findFileData.nFileSizeLow;
+    ent_.stat.size = fileSize.QuadPart;
+
+    if (findFileData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) {
+      ent_.stat.mode = _S_IFREG;
+    }
+    else if (findFileData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) {
+      ent_.stat.mode = _S_IFDIR | S_IEXEC | S_IXGRP | S_IXOTH;
+    }
+    else {
+      ent_.stat.mode = _S_IFREG;
+    }
+    if (findFileData.dwFileAttributes & FILE_ATTRIBUTE_READONLY) {
+      ent_.stat.mode |= 0444;
+    }
+    else {
+      ent_.stat.mode |= 0666;
+    }  
+
     return &ent_;
   }
+
+ public:
+  explicit WinDirHandle(const char* path, bool strict) {
+    int err = 0;
+    dirWPath_ = w_string_piece(path).asWideUNC();
+
+    h_ = Win32Handle(intptr_t(CreateFileW(
+        dirWPath_.c_str(),
+        GENERIC_READ,
+        FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE,
+        nullptr,
+        OPEN_EXISTING,
+        // Note: FILE_FLAG_OPEN_REPARSE_POINT is equivalent to O_NOFOLLOW,
+        // and FILE_FLAG_BACKUP_SEMANTICS is equivalent to O_DIRECTORY
+        (strict ? FILE_FLAG_OPEN_REPARSE_POINT : 0) |
+            FILE_FLAG_BACKUP_SEMANTICS,
+        nullptr)));
+
+    if (!h_) {
+      throw std::system_error(
+          GetLastError(),
+          std::system_category(),
+          std::string("CreateFileW for opendir: ") + path);
+    }
+
+    // Use Win7 compatibility mode for readDir()
+    if (getenv("WATCHMAN_WIN7_COMPAT") &&
+        getenv("WATCHMAN_WIN7_COMPAT")[0] == '1') {
+      win7_ = true;
+    }
+
+    memset(&ent_, 0, sizeof(ent_));
+    ent_.d_name = nameBuf_;
+    ent_.has_stat = true;
+    if (path[1] == ':') {
+      ent_.stat.dev = tolower(path[0]) - 'a';
+    }
+  }
+
+  const watchman_dir_ent* readDir() override {
+    if (win7_) {
+      return readDirWin7();
+    }
+    try {
+      return readDirWin8();
+    }
+    catch (const std::system_error& err) {
```

Comment:
Thanks for the feeback, Your are absolutely right, the fallback should be scoped, if possible, only to our case, Otherwise we could end up using both implementations on all platforms without even knowing it.